server/registry/internal/storage/list.go

```diff
@@ -57,43 +69,43 @@ func (c *Client) ListProjects(ctx context.Context, opts PageOptions) (ProjectLis
 		return ProjectList{}, err
 	}
 
-	lock()
-	var projects []models.Project
-	err = c.db.
-		Order("key").
-		Offset(token.Offset).
-		Limit(100000).
-		Find(&projects).Error
-	unlock()
-
-	if err != nil {
-		return ProjectList{}, status.Error(codes.Internal, err.Error())
-	}
-
 	response := ProjectList{
 		Projects: make([]models.Project, 0, opts.Size),
 	}
 
-	for _, project := range projects {
-		match, err := filter.Matches(projectMap(project))
+	for {
+		lock()
+		var page []models.Project
+		op := c.db.Order("key").Limit(limit(opts))
+		err := op.Offset(token.Offset).Find(&page).Error
+		unlock()
+
 		if err != nil {
-			return response, err
-		} else if !match {
-			token.Offset++
-			continue
+			return ProjectList{}, status.Error(codes.Internal, err.Error())
+		} else if len(page) == 0 {
+			break
 		}
 
-		if len(response.Projects) < int(opts.Size) {
-			response.Projects = append(response.Projects, project)
-			token.Offset++
-		} else if len(response.Projects) == int(opts.Size) {
-			response.Token, err = encodeToken(token)
+		for _, v := range page {
+			match, err := filter.Matches(projectMap(v))
 			if err != nil {
-				return response, status.Error(codes.Internal, err.Error())
+				return ProjectList{}, err
+			} else if !match {
+				token.Offset++
+				continue
+			}
+
+			if len(response.Projects) == int(opts.Size) {
+				response.Token, err = encodeToken(token)
```

Comment:
I'm unsure if this warrants a comment but it was the least-clear part to me...
// If we get here then we've read too much to fit in one page, so we return a token and everything up to this point.

Reply:
Right, it's definitely the most "clever" part of the code. I'll leave it as-is since we haven't had comments on it up to this point, and don't really want to add the same comment in so many different places. It's probably worth adding if we refactor this file into one general-purpose listing method.